panic: Fail in goroutine after TestFanoutSyncPoolE2E has completed

**Describe the bug**
After upgrading the cloudevents SDK to v2.0.0-rc4, TestFanoutSyncPoolE2E fails with the following error:
```
panic: Fail in goroutine after TestFanoutSyncPoolE2E has completed

goroutine 6762 [running]:
testing.(*common).Fail(0xc0002a8c60)
	/usr/lib/go/src/testing/testing.go:613 +0x11a
testing.(*common).Fail(0xc000c81e60)
	/usr/lib/go/src/testing/testing.go:607 +0x92
testing.(*common).Errorf(0xc000c81e60, 0xfe48e7, 0x22, 0xc0083dfc90, 0x2, 0x2)
	/usr/lib/go/src/testing/testing.go:712 +0x90
github.com/google/knative-gcp/pkg/broker/handler/pool/testing.assertEvent(0xc000c81e60, 0xc0006d0d40, 0x0, 0xc000660000, 0x61)
	/home/ian/go/src/github.com/google/knative-gcp/pkg/broker/handler/pool/testing/helper.go:493 +0x184
github.com/google/knative-gcp/pkg/broker/handler/pool/testing.(*Helper).VerifyNextTargetRetryEvent.func1(0xc000c81e60, 0xc0006d0d40, 0xc003908ee0, 0xc003699080, 0x52)
	/home/ian/go/src/github.com/google/knative-gcp/pkg/broker/handler/pool/testing/helper.go:444 +0xc5
github.com/google/knative-gcp/pkg/broker/handler/pool/testing.(*Helper).VerifyNextTargetRetryEvent(0xc0001a47c0, 0x1169820, 0xc0001a43c0, 0xc000c81e60, 0xc003699080, 0x52, 0xc0006d0d40)
	/home/ian/go/src/github.com/google/knative-gcp/pkg/broker/handler/pool/testing/helper.go:462 +0x59c
created by github.com/google/knative-gcp/pkg/broker/handler/pool.TestFanoutSyncPoolE2E.func5
	/home/ian/go/src/github.com/google/knative-gcp/pkg/broker/handler/pool/fanout_test.go:212 +0x1c4
exit status 2
FAIL	github.com/google/knative-gcp/pkg/broker/handler/pool	17.042s
```

 **Expected behavior**
The test should wait for all helpers to complete before finishing to prevent this type of error.

 **To Reproduce**

1. Unpin the cloudevents SDK and upgrade to v2.0.0-rc4.
2. Run TestFanoutSyncPoolE2E.

--- a/pkg/broker/handler/pool/fanout_test.go
+++ b/pkg/broker/handler/pool/fanout_test.go
@@ -24,6 +24,7 @@ import (
 
 	"github.com/cloudevents/sdk-go/v2/event"
 	"github.com/google/go-cmp/cmp"
+	"golang.org/x/sync/errgroup"
 
 	"github.com/google/knative-gcp/pkg/broker/config"
 	"github.com/google/knative-gcp/pkg/broker/eventutil"
@@ -134,16 +135,29 @@ func TestFanoutSyncPoolE2E(t *testing.T) {
 	t.Run("broker's targets receive fanout events", func(t *testing.T) {
 		// Set timeout context so that verification can be done before
 		// exiting test func.
-		vctx, cancel := context.WithTimeout(ctx, 2*time.Second)
+		ctx, cancel := context.WithTimeout(ctx, 2*time.Second)
 		defer cancel()
 
-		go helper.VerifyNextTargetEvent(vctx, t, t1.Key(), &e)
-		go helper.VerifyNextTargetEvent(vctx, t, t2.Key(), &e)
-		go helper.VerifyNextTargetEvent(vctx, t, t3.Key(), nil)
+		group, ctx := errgroup.WithContext(ctx)
+		group.Go(func() error {
+			helper.VerifyNextTargetEvent(ctx, t, t1.Key(), &e)
+			return nil
+		})
+		group.Go(func() error {
+			helper.VerifyNextTargetEvent(ctx, t, t2.Key(), &e)
+			return nil
+		})
+		group.Go(func() error {
+			helper.VerifyNextTargetEvent(ctx, t, t3.Key(), nil)
+			return nil
+		})
 
 		// Only send an event to broker1.
 		helper.SendEventToDecoupleQueue(ctx, t, b1.Key(), &e)
-		<-vctx.Done()
+
+		if err := group.Wait(); err != nil {
+			t.Error(err)
+		}
 	})
 
 	t.Run("target with unmatching filter didn't receive event", func(t *testing.T) {
@@ -152,72 +166,122 @@ func TestFanoutSyncPoolE2E(t *testing.T) {
 
 		// Set timeout context so that verification can be done before
 		// exiting test func.
-		vctx, cancel := context.WithTimeout(ctx, 2*time.Second)
+		ctx, cancel := context.WithTimeout(ctx, 2*time.Second)
 		defer cancel()
 
+		group, ctx := errgroup.WithContext(ctx)
 		// The old targets for broker1 should still receive the event.
-		go helper.VerifyNextTargetEvent(vctx, t, t1.Key(), &e)
-		go helper.VerifyNextTargetEvent(vctx, t, t2.Key(), &e)
+		group.Go(func() error {
+			helper.VerifyNextTargetEvent(ctx, t, t1.Key(), &e)
+			return nil
+		})
+		group.Go(func() error {
+			helper.VerifyNextTargetEvent(ctx, t, t2.Key(), &e)
+			return nil
+		})
 		// The new target for broker1 shouldn't receive the event
 		// because the event doesn't match its filter.
-		go helper.VerifyNextTargetEvent(vctx, t, t4.Key(), nil)
+		group.Go(func() error {
+			helper.VerifyNextTargetEvent(ctx, t, t4.Key(), nil)
+			return nil
+		})
 		// Target for broker2 still shouldn't receive any event.
-		go helper.VerifyNextTargetEvent(vctx, t, t3.Key(), nil)
+		group.Go(func() error {
+			helper.VerifyNextTargetEvent(ctx, t, t3.Key(), nil)
+			return nil
+		})
 
 		// Only send an event to broker1.
 		helper.SendEventToDecoupleQueue(ctx, t, b1.Key(), &e)
-		<-vctx.Done()
+
+		if err := group.Wait(); err != nil {
+			t.Error(err)
+		}
 	})
 
 	t.Run("event sent to a broker didn't reach another broker's targets", func(t *testing.T) {
 		// Set timeout context so that verification can be done before
 		// exiting test func.
-		vctx, cancel := context.WithTimeout(ctx, 2*time.Second)
+		ctx, cancel := context.WithTimeout(ctx, 2*time.Second)
 		defer cancel()
 
+		group, ctx := errgroup.WithContext(ctx)
 		// This time targets for broker1 shouldn't receive any event.
-		go helper.VerifyNextTargetEvent(vctx, t, t1.Key(), nil)
-		go helper.VerifyNextTargetEvent(vctx, t, t2.Key(), nil)
+		group.Go(func() error {
+			helper.VerifyNextTargetEvent(ctx, t, t1.Key(), nil)
+			return nil
+		})
+		group.Go(func() error {
+			helper.VerifyNextTargetEvent(ctx, t, t2.Key(), nil)
+			return nil
+		})
 		// Target for broker2 should receive the event.
-		go helper.VerifyNextTargetEvent(vctx, t, t3.Key(), &e)
+		group.Go(func() error {
+			helper.VerifyNextTargetEvent(ctx, t, t3.Key(), &e)
+			return nil
+		})
 
 		// Only send an event to broker2.
 		helper.SendEventToDecoupleQueue(ctx, t, b2.Key(), &e)
-		<-vctx.Done()
+
+		if err := group.Wait(); err != nil {
+			t.Error(err)
+		}
 	})
 
 	t.Run("event failed initial delivery was sent to retry queue", func(t *testing.T) {
-		// Set timeout context so that verification can be done before
-		// exiting test func.
-		vctx, cancel := context.WithTimeout(ctx, 2*time.Second)
-		defer cancel()
-
-		go helper.VerifyAndRespondNextTargetEvent(ctx, t, t3.Key(), &e, nil, http.StatusInternalServerError, 0)
-		go helper.VerifyNextTargetRetryEvent(ctx, t, t3.Key(), &e)
+		group, ctx := errgroup.WithContext(ctx)
+		group.Go(func() error {
+			helper.VerifyAndRespondNextTargetEvent(ctx, t, t3.Key(), &e, nil, http.StatusInternalServerError, 0)
+			return nil
+		})
+		group.Go(func() error {
+			helper.VerifyNextTargetRetryEvent(ctx, t, t3.Key(), &e)
+			return nil
+		})
 
 		helper.SendEventToDecoupleQueue(ctx, t, b2.Key(), &e)
-		<-vctx.Done()
+
+		if err := group.Wait(); err != nil {
+			t.Error(err)
+		}
 	})
 
 	t.Run("event with delivery timeout was sent to retry queue", func(t *testing.T) {
 		// Set timeout context so that verification can be done before
 		// exiting test func.
-		vctx, cancel := context.WithTimeout(ctx, 2*time.Second)
+		ctx, cancel := context.WithTimeout(ctx, 2*time.Second)
 		defer cancel()
 
+		group, ctx := errgroup.WithContext(ctx)
 		// Verify the event to t1 was received.
-		go helper.VerifyNextTargetEventAndDelayResp(ctx, t, t1.Key(), &e, time.Second)
+		group.Go(func() error {
+			helper.VerifyNextTargetEventAndDelayResp(ctx, t, t1.Key(), &e, time.Second)
+			return nil
+		})
 		// Because of the delay, t1 delivery should timeout.
 		// Thus the event should have been sent to the retry queue.
-		go helper.VerifyNextTargetRetryEvent(ctx, t, t1.Key(), &e)
+		group.Go(func() error {
+			helper.VerifyNextTargetRetryEvent(ctx, t, t1.Key(), &e)
+			return nil
+		})
 		// The same event should be received by t2.
-		go helper.VerifyNextTargetEvent(vctx, t, t2.Key(), &e)
+		group.Go(func() error {
+			helper.VerifyNextTargetEvent(ctx, t, t2.Key(), &e)
+			return nil
+		})
 		// But t2 shouldn't receive any retry event because the initial delay
 		// was successful.
-		go helper.VerifyNextTargetRetryEvent(vctx, t, t2.Key(), nil)
+		group.Go(func() error {
+			helper.VerifyNextTargetRetryEvent(ctx, t, t2.Key(), nil)
+			return nil
+		})
 
 		helper.SendEventToDecoupleQueue(ctx, t, b1.Key(), &e)
-		<-vctx.Done()
+
+		if err := group.Wait(); err != nil {
+			t.Error(err)
+		}
 	})
 
 	t.Run("event replied was sent to broker ingress", func(t *testing.T) {
@@ -232,33 +296,40 @@ func TestFanoutSyncPoolE2E(t *testing.T) {
 		// -1 because the delivery processor should decrement remaining hops.
 		eventutil.UpdateRemainingHops(ctx, &wantReply, hops-1)
 
-		// Set timeout context so that verification can be done before
-		// exiting test func.
-		vctx, cancel := context.WithTimeout(ctx, 2*time.Second)
-		defer cancel()
-
-		go helper.VerifyAndRespondNextTargetEvent(ctx, t, t3.Key(), &e, &reply, http.StatusOK, 0)
-		go helper.VerifyNextBrokerIngressEvent(ctx, t, b2.Key(), &wantReply)
+		group, ctx := errgroup.WithContext(ctx)
+		group.Go(func() error {
+			helper.VerifyAndRespondNextTargetEvent(ctx, t, t3.Key(), &e, &reply, http.StatusOK, 0)
+			return nil
+		})
+		group.Go(func() error {
+			helper.VerifyNextBrokerIngressEvent(ctx, t, b2.Key(), &wantReply)
+			return nil
+		})
 
 		helper.SendEventToDecoupleQueue(ctx, t, b2.Key(), &e)
-		<-vctx.Done()
+
+		if err := group.Wait(); err != nil {
+			t.Error(err)
+		}
 	})
 
 	t.Run("event delivered after broker decouple queue update", func(t *testing.T) {
 		helper.RenewBroker(ctx, t, b2.Key())
 		signal <- struct{}{}
 
-		// Set timeout context so that verification can be done before
-		// exiting test func.
-		vctx, cancel := context.WithTimeout(ctx, 2*time.Second)
-		defer cancel()
-
+		group, ctx := errgroup.WithContext(ctx)
 		// Target for broker2 should continue to receive the event.
-		go helper.VerifyNextTargetEvent(vctx, t, t3.Key(), &e)
+		group.Go(func() error {
+			helper.VerifyNextTargetEvent(ctx, t, t3.Key(), &e)
+			return nil
+		})
 
 		// Only send an event to broker2.
 		helper.SendEventToDecoupleQueue(ctx, t, b2.Key(), &e)
-		<-vctx.Done()
+
+		if err := group.Wait(); err != nil {
+			t.Error(err)
+		}
 	})
 }
 

--- a/pkg/broker/handler/pool/retry_test.go
+++ b/pkg/broker/handler/pool/retry_test.go
@@ -23,6 +23,7 @@ import (
 
 	"github.com/cloudevents/sdk-go/v2/event"
 	"github.com/google/go-cmp/cmp"
+	"golang.org/x/sync/errgroup"
 
 	"github.com/google/knative-gcp/pkg/broker/config"
 	"github.com/google/knative-gcp/pkg/broker/eventutil"
@@ -129,55 +130,86 @@ func TestRetrySyncPoolE2E(t *testing.T) {
 	t.Run("target with same broker but different trigger did't receive retry events", func(t *testing.T) {
 		// Set timeout context so that verification can be done before
 		// exiting test func.
-		vctx, cancel := context.WithTimeout(ctx, 2*time.Second)
+		ctx, cancel := context.WithTimeout(ctx, 2*time.Second)
 		defer cancel()
 
+		group, ctx := errgroup.WithContext(ctx)
 		// Target1 for broker1 should receive the event e1.
-		go helper.VerifyNextTargetEvent(vctx, t, t1.Key(), &e1)
-		// Target2 for broker1 should't receive the event e2.
-		go helper.VerifyNextTargetEvent(vctx, t, t2.Key(), nil)
+		group.Go(func() error {
+			helper.VerifyNextTargetEvent(ctx, t, t1.Key(), &e1)
+			return nil
+		})
+		// Target2 for broker1 shouldn't receive the event e2.
+		group.Go(func() error {
+			helper.VerifyNextTargetEvent(ctx, t, t2.Key(), nil)
+			return nil
+		})
 
 		// Only send an event to trigger topic 1.
 		helper.SendEventToRetryQueue(ctx, t, t1.Key(), &e1)
-		<-vctx.Done()
+
+		if err := group.Wait(); err != nil {
+			t.Error(err)
+		}
 	})
 
 	t.Run("target with different broker did't receive retry events", func(t *testing.T) {
 		// Set timeout context so that verification can be done before
 		// exiting test func.
-		vctx, cancel := context.WithTimeout(ctx, 2*time.Second)
+		ctx, cancel := context.WithTimeout(ctx, 2*time.Second)
 		defer cancel()
 
-		// Target1 for broker1 should't  receive the event e3.
-		go helper.VerifyNextTargetEvent(vctx, t, t1.Key(), nil)
-		// Target2 for broker1 should't receive the event e2.
-		go helper.VerifyNextTargetEvent(vctx, t, t2.Key(), nil)
+		group, ctx := errgroup.WithContext(ctx)
+		// Target1 for broker1 shouldn't  receive the event e3.
+		group.Go(func() error {
+			helper.VerifyNextTargetEvent(ctx, t, t1.Key(), nil)
+			return nil
+		})
+		// Target2 for broker1 shouldn't receive the event e2.
+		group.Go(func() error {
+			helper.VerifyNextTargetEvent(ctx, t, t2.Key(), nil)
+			return nil
+		})
 		// Target3 for broker2 should receive the event e3.
-		go helper.VerifyNextTargetEvent(vctx, t, t3.Key(), &e3)
+		group.Go(func() error {
+			helper.VerifyNextTargetEvent(ctx, t, t3.Key(), &e3)
+			return nil
+		})
 
 		// Only send an event to trigger topic 3.
 		helper.SendEventToRetryQueue(ctx, t, t3.Key(), &e3)
-		<-vctx.Done()
+
+		if err := group.Wait(); err != nil {
+			t.Error(err)
+		}
 	})
 
 	t.Run("broker's target receive correct retry events", func(t *testing.T) {
-		// Set timeout context so that verification can be done before
-		// exiting test func.
-		vctx, cancel := context.WithTimeout(ctx, 2*time.Second)
-		defer cancel()
-
+		group, ctx := errgroup.WithContext(ctx)
 		// Target1 for broker1 should receive the event e1.
-		go helper.VerifyNextTargetEvent(vctx, t, t1.Key(), &e1)
+		group.Go(func() error {
+			helper.VerifyNextTargetEvent(ctx, t, t1.Key(), &e1)
+			return nil
+		})
 		// Target2 for broker1 should receive the event e2.
-		go helper.VerifyNextTargetEvent(vctx, t, t2.Key(), &e2)
+		group.Go(func() error {
+			helper.VerifyNextTargetEvent(ctx, t, t2.Key(), &e2)
+			return nil
+		})
 		// Target3 for broker2 should receive the event e3.
-		go helper.VerifyNextTargetEvent(vctx, t, t3.Key(), &e3)
+		group.Go(func() error {
+			helper.VerifyNextTargetEvent(ctx, t, t3.Key(), &e3)
+			return nil
+		})
 
 		// Send different event to different trigger topic.
 		helper.SendEventToRetryQueue(ctx, t, t1.Key(), &e1)
 		helper.SendEventToRetryQueue(ctx, t, t2.Key(), &e2)
 		helper.SendEventToRetryQueue(ctx, t, t3.Key(), &e3)
-		<-vctx.Done()
+
+		if err := group.Wait(); err != nil {
+			t.Error(err)
+		}
 	})
 
 	t.Run("broker's target receive correct retry events with the latest filter", func(t *testing.T) {
@@ -193,31 +225,43 @@ func TestRetrySyncPoolE2E(t *testing.T) {
 			bm.UpsertTargets(t1)
 		})
 
+		group, ctx := errgroup.WithContext(ctx)
 		// Target1 for broker1 shouldn't receive the event e1, as filter attributes is updated.
-		go helper.VerifyNextTargetEvent(vctx, t, t1.Key(), nil)
+		group.Go(func() error {
+			helper.VerifyNextTargetEvent(vctx, t, t1.Key(), nil)
+			return nil
+		})
 		//  Target1 for broker1 should receive the event e1.
-		go helper.VerifyNextTargetEvent(vctx, t, t2.Key(), &e1)
+		group.Go(func() error {
+			helper.VerifyNextTargetEvent(vctx, t, t2.Key(), &e1)
+			return nil
+		})
 
 		// Send the same event to different trigger topic.
 		helper.SendEventToRetryQueue(ctx, t, t1.Key(), &e1)
 		helper.SendEventToRetryQueue(ctx, t, t2.Key(), &e1)
-		<-vctx.Done()
+
+		if err := group.Wait(); err != nil {
+			t.Error(err)
+		}
 	})
 
 	t.Run("event delivered after target retry queue update", func(t *testing.T) {
 		helper.RenewTarget(ctx, t, t2.Key())
 		signal <- struct{}{}
 
-		// Set timeout context so that verification can be done before
-		// exiting test func.
-		vctx, cancel := context.WithTimeout(ctx, 2*time.Second)
-		defer cancel()
-
+		group, ctx := errgroup.WithContext(ctx)
 		// t2 should continue to receive the event.
-		go helper.VerifyNextTargetEvent(vctx, t, t2.Key(), &e1)
+		group.Go(func() error {
+			helper.VerifyNextTargetEvent(ctx, t, t2.Key(), &e1)
+			return nil
+		})
 
 		helper.SendEventToRetryQueue(ctx, t, t2.Key(), &e1)
-		<-vctx.Done()
+
+		if err := group.Wait(); err != nil {
+			t.Error(err)
+		}
 	})
 }
 

--- a/pkg/broker/handler/pool/testing/helper.go
+++ b/pkg/broker/handler/pool/testing/helper.go
@@ -361,6 +361,9 @@ func (h *Helper) VerifyNextBrokerIngressEvent(ctx context.Context, t *testing.T,
 	msg, err := bIng.client.Receive(ctx)
 	if err != nil {
 		// In case Receive is stopped.
+		if wantEvent != nil {
+			t.Errorf("Unexpected error receiving event: %v", err)
+		}
 		return
 	}
 	msg.Finish(nil)
@@ -413,6 +416,9 @@ func (h *Helper) VerifyAndRespondNextTargetEvent(ctx context.Context, t *testing
 	msg, respFn, err := consumer.client.Respond(ctx)
 	if err != nil {
 		// In case Receive is stopped.
+		if wantEvent != nil {
+			t.Errorf("Unexpected error receiving event: %v", err)
+		}
 		return
 	}
 	gotEvent, err = binding.ToEvent(ctx, msg)
@@ -463,6 +469,9 @@ func (h *Helper) VerifyNextTargetRetryEvent(ctx context.Context, t *testing.T, t
 	msg, err := psTmp.Receive(ctx)
 	if err != nil {
 		// In case Receive is stopped.
+		if wantEvent != nil {
+			t.Errorf("Unexpected error receiving event: %v", err)
+		}
 		return
 	}
 	msg.Finish(nil)

